browser/go/providerwrapper/providerwrapper.go

```diff
@@ -0,0 +1,183 @@
+		}
+		payload["params"] = convertedParams
+	}
+
+	// Set up the callback function
+	resultChan := make(chan js.Value, 1)
+	errChan := make(chan error, 1)
+	var callback js.Func
+	callback = js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		defer callback.Release()
+		go func() {
+			if len(args) == 0 {
+				errChan <- errors.New("JSONRPCErrorCallback called with no arguments")
+				return
+			}
+			jsError := args[0]
+			if !jsutil.IsNullOrUndefined(jsError) {
+				errChan <- js.Error{
+					Value: jsError,
+				}
+				return
+			}
+			if len(args) < 2 {
+				errChan <- errors.New("JSONRPCErrorCallback called with null/undefined error but no results")
+				return
+			}
+			resultChan <- args[1]
+			return
+		}()
+		return nil
+	})
+
+	// Call sendAsync and use select to wait for the results.
+	c.provider.Call("sendAsync", payload, callback)
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errChan:
+		return err
+	case jsResult := <-resultChan:
+		if rpcErr := jsResult.Get("error"); !jsutil.IsNullOrUndefined(rpcErr) {
+			return jsErrorToRPCError(rpcErr)
+		}
+		if result == nil {
+			return nil
+		}
+		if err := jsutil.InefficientlyConvertFromJS(jsResult.Get("result"), result); err != nil {
+			return fmt.Errorf("could not decode JSON RPC response: %s", err.Error())
+		}
+		return nil
+	}
+}
+
+// BatchCall sends all given requests as a single batch and waits for the server
+// to return a response for all of them. The wait duration is bounded by the
+// context's deadline.
+//
+// In contrast to CallContext, BatchCallContext only returns errors that have occurred
+// while sending the request. Any error specific to a request is reported through the
+// Error field of the corresponding BatchElem.
+//
+// Note that batch calls may not be executed atomically on the server side.
+func (c *RPCClient) BatchCallContext(ctx context.Context, b []rpc.BatchElem) error {
+	log.WithField("batch", b).Error("BatchCallContext was unexpectedly called in the browser")
+	return errors.New("BatchCallContext not yet implemented")
+}
+
+// EthSubscribe registers a subscripion under the "eth" namespace.
+func (c *RPCClient) EthSubscribe(ctx context.Context, channel interface{}, args ...interface{}) (*rpc.ClientSubscription, error) {
+	log.WithField("args", args).Error("EthSubscribe was unexpectedly called in the browser")
+	return nil, errors.New("EthSubscribe not yet implemented")
+}
+
+// Close is a no-op in this implementation.
+func (c *RPCClient) Close() {
+	// no-op for now.
+}
+
+var _ rpc.Error = &rpcError{}
+
+type rpcError struct {
+	message string
+	code    int
+}
+
+func (e rpcError) Message() string {
+	return e.message
+}
+
+func (e rpcError) Error() string {
+	return fmt.Sprintf("Web3Provider returned RPC error (code %d): %s", e.code, e.message)
+}
+
+func (e rpcError) ErrorCode() int {
+	return e.code
+}
+
+func jsErrorToRPCError(jsError js.Value) rpc.Error {
```

Comment:
@fabioberger do we ever check the message, code, or type of error in block watcher or order watcher? Want to make sure that this is compatible with the rest of our code.

Reply:
We do not, but we do include them in logs sometimes. Why would they not be compatible?

Reply:
Can you point me to the code where we log these?

Reply:
The `go-ethereum` `rpc.Client` checks the JSON-RPC response and if it contains the `Error` key, it returns it's contents as the error from the call. This error object gets logged as in: https://github.com/0xProject/0x-mesh/blob/master/ethereum/blockwatch/block_watcher.go#L174

Reply:
👍 I found the concrete implementation in go-ethereum and made our implementation more closely match it. I think we should be good.